diff --git a/files/systemd/services/distro/webos/surface-manager.service b/files/systemd/services/distro/webos/surface-manager.service
index b0df8c4..f7ecdfa 100644
--- a/files/systemd/services/distro/webos/surface-manager.service
+++ b/files/systemd/services/distro/webos/surface-manager.service
@@ -20,6 +20,8 @@ Requires=ls-hubd.service
 After=ls-hubd.service
 Wants=webapp-mgr.service
 Before=webapp-mgr.service
+StartLimitBurst=10
+StartLimitIntervalSec=1

 [Service]
 Type=notify

